[COT] Add database transactions support for data synchronization

Konamiman opened the pull request:
### All Submissions:

-   [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [ ] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Add optional usage of database transactions, with a selector for the isolation level, when synchronizing data from the posts table to the custom order tables. This is controlled by two settings accessible via Woo settings - Advanced - Custom data stores.

![image](https://user-images.githubusercontent.com/937723/171148361-1e9c99ac-90a5-4630-82fb-5410050da913.png)

Additionally, this pull request makes a couple of changes to the syncrhonization code itself:

* The code that retrieves and updates existing data tries to filter out the records that don't need updated (because they haven't changed). This only works when the core column mapping is complete enough to compare full records, though.
* The "X records were expected to be processed but only Y were" error message has been removed since [the affected rows count gets really tricky when using ON DUPLICATE KEY UPDATE](https://dev.mysql.com/doc/refman/8.0/en/insert-on-duplicate.html).

Closes #33187 .

### How to test the changes in this Pull Request:

First, enable the feature by adding the following code snippet somewhere, for example at the end of `woocommerce.php`:

```php
wc_get_container()
   ->get('Automattic\WooCommerce\Internal\DataStores\Orders\CustomOrdersTableController')
   ->show_feature();
```

Then go to the related settings page, enable transactions and try to migrate some orders. You can do so via command line:

```bash
wp eval 'wc_get_container()->get(Automattic\WooCommerce\Database\Migrations\CustomOrderTable\PostsToOrdersMigrationController::class)->migrate_orders([order_id, order_id...]);'
```

To test the failing case (no data is migrated at all in case of error) you need to force an error in the middle of the migration process, when some data has already been written to the database (which will be deleted via transaction rollback). One way to achieve this is to temporarily add this method to the `PostToOrderOpTableMigrator` class:

```php
protected function process_migration_batch_for_ids_core( array $entity_ids ): void {
	throw new \Exception( 'Ops!!' );
}
```

With this in place, at the end of the process data will still have been written to the `wp_wc_orders` table when transactions are disabled, but not when they are enabled.

### Other information:

-   [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you successfully run tests with your changes locally?
-   [ ] Have you created a changelog file for each project being changed, ie `pnpm nx changelog <project>`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

Konamiman committed:
Add optional usage of database transactions for orders sync.

Includes a new setting in Advanced - Custom data stores to enable
or disable transactions usage and to select the transaction
isolation level.

Konamiman committed:
Improve the selection of data to be updated when syncing orders.

The SQL query that chooses records to be migrated to core tables
now explicitly excludes unchanged records based on the column mappings.

Also when syncing meta tables with single meta values, records for
which the value hasn't changed are now excluded from the update.

Konamiman committed:
Merge branch 'trunk' into add/db-transaction-support-for-cot-sync

botwoo commented:
:bar_chart: Test reports for this pull request have been published and are accessible through the following links:

- [REST API Test Report](https://woocommerce.github.io/woocommerce-test-reports/pr/33255/api/)
- E2E Test Report (TBC)

Latest commit referenced in the reports: `Merge branch 'trunk' into add/db-transaction-support-for-cot-sync` 96b5462127cb24abb02d5bbda8d52ce8f76cf29f
This comment will automatically be updated with the latest referenced commit when you push new changes to this pull request.

---

_Visit the [WooCommerce Test Reports](https://woocommerce.github.io/woocommerce-test-reports) homepage to view all published reports. See the [FAQs](https://github.com/woocommerce/woocommerce-test-reports/wiki/FAQs) page if you're having problems accessing them._

Konamiman committed:
Fix MetaToCustomTableMigrator to work with order addresses table.

Konamiman committed:
Add unit tests for db transactions on PostsToOrdersMigrationController

Also add the DynamicDecorator::call_original_method method

Konamiman committed:
Add changelog file

barryhughes commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/PostsToOrdersMigrationController.php:
```diff
@@ -71,9 +72,101 @@ public function get_migrated_meta_keys() {
+		}
+
+		if ( ! $errors_were_logged && $using_transactions ) {
+			$this->commit_transaction();
```
Do we need to take any action (whether re-running/scheduling a re-run of the migration step, or simply logging the failure) in the event that the commit fails?

Konamiman replied:
The failure is laready being logged, see the `db_query` method in the same class. As for re-scheduling, that's important indeed, but the robustness of the entire sync process is pending work to be done in a separate PR.

barryhughes commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/PostsToOrdersMigrationController.php:
```diff
@@ -71,9 +72,101 @@ public function get_migrated_meta_keys() {
-			$this->do_orders_migration_step( $migrator, $order_post_ids );
+			$errors_were_logged = $this->do_orders_migration_step( $migrator, $order_post_ids );
+			if ( $errors_were_logged && $using_transactions ) {
+				$this->rollback_transaction();
```
Should we take any action if the rollback fails? I'm sure that would be unusual, but it's perhaps possible?

Konamiman replied:
It's my understanding that any transaction that is not explicitly commited gets rolled back sooner or later, either explicitly by a `ROLLBACK` command, or when the connection is closed. Any failure here will be logged though, I think that should be enough.

barryhughes commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/PostToOrderAddressTableMigrator.php:
```diff
@@ -148,48 +148,13 @@ public function get_meta_column_config(): array {
+	 * @param array $entity_ids The ids of the entities being inserted or updated.
+	 * @return string The additional string for the WHERE clause.
 	 */
-	protected function get_already_migrated_records( array $entity_ids ): array {
```
Noting we're not in the internal namespace here and that neither the class or method is marked `@internal` ... is this method safe to remove without any notice? It seems like it was present [in the last release](https://github.com/woocommerce/woocommerce/blob/6.5.1/plugins/woocommerce/src/Database/Migrations/CustomOrderTable/WPPostToOrderAddressTableMigrator.php#L161) as a public method. 

Perhaps this was already discussed in an earlier change? It may not be too big a deal at this point since we're dealing with code that lives behind a feature flag, etc, but I wanted to flag it.

Konamiman replied:
Yeah I would have preferred to have put all this code in the `Internal` namespace , but anyway, as you say it's not a big deal while the feature is in development.

barryhughes commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/PostToOrderAddressTableMigrator.php:
```diff
@@ -148,48 +148,13 @@ public function get_meta_column_config(): array {
-		);
-
-		return array_column( $already_migrated_entity_ids, null, 'source_id' );
+	protected function get_additional_where_clause_for_get_data_to_insert_or_update( array $entity_ids ): string {
```
Is the array of `$entity_ids` needed/expected to be needed in the future?

Konamiman replied:
We don't really know how the method will be used in the future, so it's better to supply it with all the context information that might be relevant (adding it later would be complicated); any implementation not needing this data can simply discard it, as it's the case here.

barryhughes commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/PostsToOrdersMigrationController.php:
```diff
@@ -71,9 +72,101 @@ public function get_migrated_meta_keys() {
+		$transaction_isolation_level             = get_option( CustomOrdersTableController::DB_TRANSACTIONS_ISOLATION_LEVEL_OPTION, CustomOrdersTableController::DEFAULT_DB_TRANSACTIONS_ISOLATION_LEVEL );
+		$set_transaction_isolation_level_command = "SET TRANSACTION ISOLATION LEVEL $transaction_isolation_level";
+
+		if ( ! $this->db_query( $set_transaction_isolation_level_command ) ) {
```
The linter isn't flagging it because we're using our own method instead of `$wpdb->query()` directly, but otherwise this would trigger:

`Use placeholders and $wpdb->prepare(); found $set_transaction_isolation_level_command`

I guess in this case we can't pass the isolation level string through `$wpdb->prepare()`, but what about comparing it against a set of allowed values (or even escaping it via a preg_replace to strip anything but alphabetical and whitespace chars)?

barryhughes commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/PostsToOrdersMigrationController.php:
```diff
@@ -71,9 +72,101 @@ public function get_migrated_meta_keys() {
+		}
+
+		$transaction_isolation_level             = get_option( CustomOrdersTableController::DB_TRANSACTIONS_ISOLATION_LEVEL_OPTION, CustomOrdersTableController::DEFAULT_DB_TRANSACTIONS_ISOLATION_LEVEL );
+		$set_transaction_isolation_level_command = "SET TRANSACTION ISOLATION LEVEL $transaction_isolation_level";
```
I *think* this may be fine, given the context and given what MySQL even allows for this sort of statement, but I wanted to pause and ask anyway: is there any other escaping we might do here for safety? Except for the fact we're using our own `db_query()` method, this would be normally have been flagged by the linter.

For instance, could we compare against a list of allowed values (such as if we had an 'enumeration' of [these strings](https://github.com/woocommerce/woocommerce/pull/33255/files#diff-b8eca23600b4f2aa74ae748c318e9a40316428531b1fcf4fefe03e26e348b4f4R408-R411))?

Konamiman replied:
That's a good point, we are relying on an option that could have an arbitrary value. I'll think of a way to sanitize it.

barryhughes commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/PostsToOrdersMigrationController.php:
```diff
@@ -71,9 +72,101 @@ public function get_migrated_meta_keys() {
+		}
+
+		$transaction_isolation_level             = get_option( CustomOrdersTableController::DB_TRANSACTIONS_ISOLATION_LEVEL_OPTION, CustomOrdersTableController::DEFAULT_DB_TRANSACTIONS_ISOLATION_LEVEL );
+		$set_transaction_isolation_level_command = "SET TRANSACTION ISOLATION LEVEL $transaction_isolation_level";
```
I think this is fine, given the context and given what MySQL even allows for this sort of statement, but I wanted to pause and ask anyway: is there any other escaping we might do here for safety (except for the fact we're using our own `db_query()` method, this would be normally have been flagged by the linter)?

Possibly comparing the string against a list of allowed values? (May not be necessary or even practical do to so, if which case we can move on.)

barryhughes commented:
Test fail to look at:

```
There was 1 failure:

1) PostsToOrdersMigrationControllerTest::test_db_transaction_completes_if_configured_and_no_errors
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
     0 => 'SET TRANSACTION ISOLATION LEV..._LEVEL'
     1 => 'START TRANSACTION'
-    2 => 'COMMIT'
+    2 => 'ROLLBACK'
 )
```

Konamiman commented:
> Test fail to look at:

Weird, it's working for me locally 😕 @barryhughes are you able to reproduce the fail yourself?

Konamiman committed:
Add validation for the transaction isolation level name.

The name is retrieved from an option so it's compared against
a list of allowed values before being used.

barryhughes reviewed (approved):
Works well; my operational order data table was actually slightly out-of-date when I started testing, so happily that let me observe the rollback-upon-error.

barryhughes commented:
E2E fail which I suspect is unrelated. Re-running to be safe (otherwise this is good to merge).

Konamiman commented:
@barryhughes The failures in E2E tests seem unrelated and are apparently failing on other PRs too, so I'm merging.
